scripts/kb.py

```diff
@@ -0,0 +1,633 @@
+                language=self._language, qids=qids, db_conn=self._db_conn
+            )
+
+            # Assemble descriptions to be embedded.
+            ent_descs = [
+                " ".join({entities[qid].name, *entities[qid].aliases})
+                + " "
+                + (
+                    entities[qid].description
+                    if entities[qid].description
+                    else (
+                        entities[qid].article_text[:500]
+                        if entities[qid].article_text
+                        else ""
+                    )
+                )
+                for qid in qids
+            ]
+
+            for row_id_offset, qid, desc_vector in zip(
+                range(len(qids)),
+                qids,
+                [
+                    ent_desc_doc.vector
+                    for ent_desc_doc in nlp.pipe(texts=ent_descs, n_process=n_jobs)
+                ],
+            ):
+                self._annoy.add_item(
+                    # Annoy expects index to start with 0, so we index each entities vector by its entities.ROWID value
+                    # in the database shifted by -1.
+                    row_id + row_id_offset - 1,
+                    desc_vector
+                    if isinstance(desc_vector, numpy.ndarray)
+                    else desc_vector.get(),
+                )
+
+        logger.info("Building ANN index.")
+        self._annoy.build(n_trees=self._n_trees, n_jobs=n_jobs)
+
+    def get_candidates_all(
+        self, mentions: Iterator[Iterable[Span]]
+    ) -> Iterator[Iterable[Iterable[Candidate]]]:
+        """
+        Retrieve candidate entities for specified mentions per document. If no candidate is found for a given mention,
+        an empty list is returned.
+        mentions (Iterator[Iterable[Span]]): Mentions per documents for which to get candidates.
+        YIELDS (Iterator[Iterable[Iterable[Candidate]]]): Identified candidates per document.
+        """
+        for mentions_in_doc in mentions:
+            mentions_in_doc = tuple(mentions_in_doc)
+            alias_matches = self._fetch_candidates_by_alias(mentions_in_doc)
+            fts_matches = self._fetch_candidates_by_fts(mentions_in_doc)
+            # Candidates for each mention per document.
+            candidates: List[List[Candidate]] = []
+
+            for i, mention in enumerate(mentions_in_doc):
+                candidates.append([])
+
+                for cand_data in alias_matches.get(mention.text, []):
+                    candidates[i].append(
+                        Candidate(
+                            kb=self,
+                            entity_freq=cand_data["sum_occurence_count"],
+                            prior_prob=cand_data["max_prior_prob"],
+                            entity_vector=next(
+                                iter(self._get_vectors([cand_data["rowid"]]))
+                            ),
+                            # Hashes aren't used by WikiKB.
```

Comment:
This clarifies again that the current implementations are too much interdependent

Reply:
We could do some additional refactoring (in https://github.com/explosion/spaCy/pull/11669) and have:
- a base `Candidate` class with properties `kb`, `entity_vector`, `prior_prob` - I think these attributes we want for all KB implementations - and
- a backwards-compatible `InMemoryLookupKBCandidate` (ideally with a name that's less clunky) that adds `entity_freq` and `alias_hash`.

The base candidate class would be sufficient for `WikiKB`. I didn't do this yet because it's not necessary for this to work. I could do that though?

Reply:
I wouldn't necessarily block this PR on that refactor, but the refactor does sound like a good idea, yea! Then again, maybe we want to continue with some further experiments first like the learnable CG+EL system, which might make all of these discussions obsolete again... so maybe let's just add this to the backlog.

Reply:
Agreed. Added a `Candidate` refactoring issue to the EL backlog.